File: dart_core/test/webhook_test.dart
import 'package:telebirr_sdk_core/src/webhook.dart';
import 'package:test/test.dart';

void main() {
  group('NotificationTradeStatus', () {
    test('has correct values', () {
      expect(NotificationTradeStatus.paying.value, 'Paying');
      expect(NotificationTradeStatus.expired.value, 'Expired');
      expect(NotificationTradeStatus.pending.value, 'Pending');
      expect(NotificationTradeStatus.completed.value, 'Completed');
      expect(NotificationTradeStatus.failure.value, 'Failure');
    });
  });

  group('buildNotificationSignString', () {
    test('excludes sign and sign_type', () {
      final payload = {
        'merch_order_id': 'order-123',
        'payment_order_id': 'pay-456',
        'trade_status': 'Completed',
        'sign': 'some-signature',
        'sign_type': 'SHA256WithRSA',
      };

      final result = buildNotificationSignString(payload);

      expect(result, contains('merch_order_id=order-123'));
      expect(result, contains('payment_order_id=pay-456'));
      expect(result, contains('trade_status=Completed'));
      expect(result, isNot(contains('sign=')));
      expect(result, isNot(contains('sign_type=')));
    });

    test('sorts keys lexicographically', () {
      final payload = {
        'zebra': 'z',
        'alpha': 'a',
        'middle': 'm',
      };

      final result = buildNotificationSignString(payload);
      expect(result, 'alpha=a&middle=m&zebra=z');
    });
  });

  group('NotificationPayload', () {
    test('accesses fields correctly', () {
      final raw = {
        'merch_order_id': 'order-123',
        'payment_order_id': 'pay-456',
        'trade_status': 'Completed',
        'total_amount': '100',
      };

      final payload = NotificationPayload(raw);

      expect(payload.merchOrderId, 'order-123');
      expect(payload.paymentOrderId, 'pay-456');
      expect(payload.tradeStatus, 'Completed');
      expect(payload.totalAmount, '100');
    });
  });

  group('verifyNotification', () {
    test('returns false when sign is missing', () {
      final payload = NotificationPayload({
        'merch_order_id': 'order-123',
        'trade_status': 'Completed',
      });

      expect(verifyNotification(payload, 'fake-public-key'), isFalse);
    });
  });
}

File: ts-core/tests/webhook.test.ts
import { verifyNotification, NotificationTradeStatus } from '../src/webhook.js';
import { generateKeyPairSync, createSign } from 'node:crypto';

let privateKey = '';
let publicKey = '';

beforeAll(() => {
  const { privateKey: priv, publicKey: pub } = generateKeyPairSync('rsa', {
    modulusLength: 2048,
    publicKeyEncoding: { type: 'spki', format: 'pem' },
    privateKeyEncoding: { type: 'pkcs8', format: 'pem' },
  });
  privateKey = priv;
  publicKey = pub;
});

describe('NotificationTradeStatus', () => {
  it('has correct values', () => {
    expect(NotificationTradeStatus.PAYING).toBe('Paying');
    expect(NotificationTradeStatus.EXPIRED).toBe('Expired');
    expect(NotificationTradeStatus.PENDING).toBe('Pending');
    expect(NotificationTradeStatus.COMPLETED).toBe('Completed');
    expect(NotificationTradeStatus.FAILURE).toBe('Failure');
  });
});

describe('verifyNotification', () => {
  it('returns true for valid signature', () => {
    const payload = {
      merch_order_id: 'order-123',
      payment_order_id: 'pay-456',
      trade_status: NotificationTradeStatus.COMPLETED,
      trans_end_time: '20240101120000',
      sign: '',
      sign_type: 'SHA256WithRSA',
    };

    const signString = Object.keys(payload)
      .filter((k) => k !== 'sign' && k !== 'sign_type')
      .sort()
      .map((k) => `${k}=${payload[k as keyof typeof payload]}`)
      .join('&');

    const sign = createSign('SHA256');
    sign.update(signString);
    sign.end();
    const signature = sign.sign(privateKey, 'base64');

    const signedPayload = { ...payload, sign: signature };
    expect(verifyNotification(signedPayload, publicKey)).toBe(true);
  });

  it('returns false for invalid signature', () => {
    const payload = {
      merch_order_id: 'order-123',
      payment_order_id: 'pay-456',
      trade_status: NotificationTradeStatus.COMPLETED,
      trans_end_time: '20240101120000',
      sign: 'invalid-signature',
      sign_type: 'SHA256WithRSA',
    };

    expect(verifyNotification(payload, publicKey)).toBe(false);
  });

  it('returns false when sign is missing', () => {
    const payload = {
      merch_order_id: 'order-123',
      payment_order_id: 'pay-456',
      trade_status: NotificationTradeStatus.COMPLETED,
      trans_end_time: '20240101120000',
      sign_type: 'SHA256WithRSA',
    };

    expect(verifyNotification(payload, publicKey)).toBe(false);
  });
});
